### PR TITLE
drivers: versal: mailbox communication

### PR DESCRIPTION
**Pull request by ldts**
Mailbox driver to communicate with the PLM firmware executing on the
Microblaze processor.

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

reference: https://github.com/OP-TEE/optee_os/pull/5426

### CODE REVIEW

File: core/drivers/versal_mbox.c [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+		}
+
+		if (!IS_ALIGNED(cmd->ibuf[i].mem.alloc_len, CACHELINE_LEN)) {
+			EMSG("length not aligned: buffer %ld - %ld",
+			     i, cmd->ibuf[i].mem.alloc_len);
+			return TEE_ERROR_GENERIC;
+		}
+
+		cache_operation(TEE_CACHEFLUSH,
+				cmd->ibuf[i].mem.buf,
```
Comment (jforissier) [inline core/drivers/versal_mbox.c:222]: Fits on the previous line I think

File: core/include/drivers/versal_mbox.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+struct ipi_buf {
+	struct versal_mbox_mem mem;
+	bool only_cache;
+};
+
+struct ipi_cmd {
+	uint32_t data[8];
+	struct ipi_buf ibuf[MAX_IPI_BUF];
+};
```
Comment (jforissier) [inline core/include/drivers/versal_mbox.h:31]: And maybe same for these structs? `versal_*`?

**Review by jforissier — COMMENTED** [review]
Hi @ldts,

A few comments, mostly cosmetic.

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/drivers/versal_mbox.c (hunk 1 above)
Comment (ldts) [reply core/drivers/versal_mbox.c:222]: I would have preferred the way it was (personal preference for readability) but ok

File: core/include/drivers/versal_mbox.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ */
+#ifndef __DRIVERS_VERSAL_MBOX_H
+#define __DRIVERS_VERSAL_MBOX_H
+
+#include <platform_config.h>
+#include <tee_api_types.h>
+#include <util.h>
+
+#define SRC 0
+#define DST 1
```
[the remark this answers is not in the document]
Comment (ldts) [reply core/include/drivers/versal_mbox.h:13]: will drop

File: core/drivers/versal_mbox.c [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+ * Jorge Ramirez-Ortiz <jorge@foundries.io>
+ */
+
+#include <initcall.h>
+#include <kernel/delay.h>
+#include <kernel/panic.h>
+#include <mm/core_mmu.h>
+#include <string.h>
+#include <tee/cache.h>
+#include "drivers/versal_mbox.h"
```
[the remark this answers is not in the document]
Comment (ldts) [reply core/drivers/versal_mbox.c:13]: unfortunately the include is not on the same directory any more so we have to keep it this way

**Comment by ldts** [comment]
tag added to commit b092b5f6e63b5cbbdcb8a14d90e0ed76f4112761

**Comment by ldts** [comment]
All the cryptographic and secure boot support that I need to submit is pending behind this pull request.

In the meantime, users requiring access  - those attending the AMD/Xilinx Security Working Group [conference](https://web.cvent.com/event/ab0ad922-94b5-4d9e-bfc6-ed5e38f6f3e8/summary) - can use the recently released Foundries.io LmP v88

Reference docs:
https://docs.foundries.io/latest/reference-manual/boards/versal.html
https://docs.foundries.io/latest/reference-manual/security/tee-on-versal-acap.html 

Reference OP-TEE (branch 3.18)
https://github.com/foundriesio/optee_os

**Comment by jforissier** [comment]
@ldts please add my Ack to the first commit and rebase to resolve the conflicts, then this will be ready for merging. Thanks!

File: core/drivers/versal_mbox.c [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+			val = nvm_id[api];
+
+		break;
+	case 12:
+		if (api < ARRAY_SIZE(puf_id))
+			val = puf_id[api];
+
+		break;
+	default:
+		val = "Invalid";
```
Comment (etienne-lms) [inline core/drivers/versal_mbox.c:172]: could `break` and leave `val` to its default `NULL` value.

File: core/drivers/versal_mbox.c [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+	}
+
+	IMSG("--- mbox: service: %s\t call: %s", module[mid],
```
Comment (etienne-lms) [inline core/drivers/versal_mbox.c:175]: I suggest to replace `--- mbox:` with `Versal mbox:`

File: core/drivers/versal_mbox.c [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+{
+	size_t i = 0;
+
+	for (i = 0; i < VERSAL_MAX_IPI_BUF; i++) {
+		if (!cmd->ibuf[i].mem.buf)
+			continue;
+
+		if (!IS_ALIGNED((uintptr_t)cmd->ibuf[i].mem.buf,
+				CACHELINE_LEN)) {
+			EMSG("address not aligned: buffer %ld - 0x%p", i,
```
Comment (etienne-lms) [inline core/drivers/versal_mbox.c:210]: s/`%ld`/`%zu`/
s/`0x%p`/`%p`/
Ditto in `versal_mbox_read_rsp()`

File: core/drivers/versal_mbox.c [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+			     cmd->ibuf[i].mem.buf);
+			return TEE_ERROR_GENERIC;
+		}
+
+		if (!IS_ALIGNED(cmd->ibuf[i].mem.alloc_len, CACHELINE_LEN)) {
+			EMSG("length not aligned: buffer %ld - %ld",
```
Comment (etienne-lms) [inline core/drivers/versal_mbox.c:216]: s/`%ld - %ld`/` %zu - %zu`/
Ditto in `versal_mbox_read_rsp()`

File: core/drivers/versal_mbox.c [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+{
+	size_t i = 0;
+
+	/* Cache operation on the IPI buffer is safe */
+	cache_operation(TEE_CACHEINVALIDATE, ipi.rsp, sizeof(rsp->data));
+
+	*status = *(uint32_t *)ipi.rsp;
+
+	if (rsp)
+		memcpy(rsp->data, ipi.rsp, sizeof(rsp->data));
```
Comment (etienne-lms) [inline core/drivers/versal_mbox.c:245]: should this move after line 248 ?

File: core/drivers/versal_mbox.c [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+
+	if (remote_status) {
+		if (err)
+			*err = remote_status;
+		/*
+		 * Check the remote code (FSBL repository) in xplmi_status.h
+		 * and the relevant service error (ie, xsecure_error.h) for
+		 * detailed information.
+		 */
+		DMSG("PLM: plm status = 0x%x, lib_status = 0x%x",
```
Comment (etienne-lms) [inline core/drivers/versal_mbox.c:328]: s/`%x`/`%"PRIx32"`/g

File: core/drivers/versal_mbox.c (hunk 6 above)
Comment (ldts) [reply core/drivers/versal_mbox.c:175]: the trace this way is clear and easy to read (characters at the start of the trace will make it more difficult to stand out). I'd prefer to keep it as is.

**Comment by ldts** [comment]
@etienne-lms @jforissier tags updated

**Comment by ldts** [comment]
@jforissier any other comments? I'd like to send some of the crypto stuff before the weekend and it wont build unless this is in

**Comment by jforissier** [comment]
@ldts good with me, @etienne-lms do you want to give an Ack/Review tag?

File: core/drivers/versal_mbox.c (hunk 6 above)
Comment (etienne-lms) [reply core/drivers/versal_mbox.c:175]: I'm not convinced it's the right way but ok.

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by ldts** [comment]
thanks @etienne-lms . 
I did put some work on the trace to make sure it is always readable and easy to inspect and being the mbox the separation between processors it did really help me debugging certain issues. I think it will benefit everyone in the end...

**Pull request merged**